refactor to let caller manage parallelism

DavidVorick opened the pull request:
I think it's going a lot faster now. Seems that single-caller performance is up vs. the old one, and multi-threaded performance is very much up. And now, the caller has full control over the threading, there's no buffer, and there's no background threading.

Overall, a substantial improvement I think.

DavidVorick commented:
*my machine is under load, so the benchmarks aren't reliable. Would be great @lukechampine if you could post the updated benchmarks from your machine.

DavidVorick commented:
Three major changes are responsible for performance increases:

1. use blake2b.Sum512
2. No locking or channels at all, only atomics
3. Every hashing operating happens in the calling thread, which means no scheduler pressure other than what threads already exist, and also threads aren't waiting on eachother to gain access to the global data / hasher.

lukechampine commented on fastrand.go:
```diff
@@ -20,65 +20,62 @@ import (
+	// with the check it is a secure 2^128 iterations.
+	if progress > 1<<63 {
+		atomic.AddUint64((*uint64)(unsafe.Pointer(&r.entropy[0])), 1)
+		atomic.StoreUint64(&r.progress, 0)
```
is reseting to zero necessary? Why not just let it overflow?
```go
if progress == 0 {
     atomic.AddUint64((*uint64)(unsafe.Pointer(&r.entropy[0])), 1)
}
```
Also, I think there may be a race on `r.entropy`, since you write to it using `atomic` but don't read using `atomic`.

lukechampine replied:
However, the race detector doesn't complain, and the race would be benign

lukechampine replied:
Confirmed a data race by changing the branch to `if progress > 1`:

```
WARNING: DATA RACE
Write at 0x00c42007c8a8 by goroutine 24:
  sync/atomic.AddInt64()
      /usr/local/go/src/runtime/race_amd64.s:276 +0xb
  github.com/NebulousLabs/fastrand.(*randReader).Read()
      /home/luke/go/src/github.com/NebulousLabs/fastrand/fastrand.go:54 +0x3ea

Previous read at 0x00c42007c8a8 by goroutine 26:
  runtime.slicecopy()
      /usr/local/go/src/runtime/slice.go:160 +0x0
  github.com/NebulousLabs/fastrand.(*randReader).Read()
      /home/luke/go/src/github.com/NebulousLabs/fastrand/fastrand.go:61 +0x201
```

The solution is to use `atomic.LoadUint64` on the first 8 bytes being read from `r.entropy`. `copy` can be used for the rest because no one ever writes to them. It's a bit ugly but I've confirmed that it resolves the race:

```go
seed := make([]byte, 40)
binary.LittleEndian.PutUint64(seed[:8], progress)
binary.LittleEndian.PutUint64(seed[8:16], atomic.LoadUint64((*uint64)(unsafe.Pointer(&r.entropy[0]))))
copy(seed[16:], r.entropy[8:])
```

DavidVorick replied:
It triggers at 1 << 63, a lot lower than the max. You need to trigger it early because you don't know how many callers are using it simultaneously, and you must be certain that no caller ever ends up with the same entropy as a previous.

I think that the race condition is benign because the early 64bits are guaranteed to be unique per thread. The only risk is that you read a value that was set a while ago and it's paired with an early 64bits that have also been used with that value.

If we're going to bother rolling over on the 128 bits I guess we should stomp the race condition too.

lukechampine commented:
Updated benchmarks:
```
BenchmarkRead32-4                     	10000000	       173 ns/op	 184.40 MB/s
BenchmarkRead512K-4                   	    1000	   1348276 ns/op	 379.74 MB/s
BenchmarkRead4Threads-4               	 3000000	       375 ns/op	 340.60 MB/s
BenchmarkRead4Threads512k-4           	    1000	   1804969 ns/op	1134.65 MB/s
BenchmarkRead64Threads-4              	  300000	      5002 ns/op	 409.38 MB/s
BenchmarkRead64Threads64k-4           	      50	  28869659 ns/op	1135.03 MB/s
BenchmarkReadCrypto32-4               	  500000	      2696 ns/op	  11.87 MB/s
BenchmarkReadCrypto4Threads32-4       	  200000	      7813 ns/op	  16.38 MB/s
BenchmarkReadCrypto4Threads512kb-4    	      20	  98890836 ns/op	  20.71 MB/s
BenchmarkReadCrypto64Threads32-4      	   10000	    119777 ns/op	  17.10 MB/s
BenchmarkReadCrypto64Threads512kb-4   	       1	1535685645 ns/op	  21.34 MB/s
BenchmarkReadCrypto512K-4             	      50	  32971840 ns/op	  15.53 MB/s
BenchmarkReadMath32-4                 	20000000	        78.9 ns/op	 405.80 MB/s
BenchmarkReadMath512K-4               	    2000	    856899 ns/op	 597.50 MB/s
```

lukechampine commented on fastrand.go:
```diff
@@ -20,65 +20,62 @@ import (
+
+	// Copy the entropy into a separately allocated slice.
+	seed := make([]byte, 40)
+	*(*uint64)(unsafe.Pointer(&seed[0])) = progress
```
I might use `binary.LittleEndian.PutUint64(seed[:8], progress)` instead here to make it more clear that you are copying bytes into the seed. (Performance is identical.)

lukechampine commented on fastrand.go:
```diff
@@ -20,65 +20,62 @@ import (
+		// bytes are part of what makes the seed unique to other threads. That
+		// leaves still a full 16 bytes of entropy that kept from the original
+		// read, more than enough to provide secure output.
+		*(*uint64)(unsafe.Pointer(&seed[16]))++
```
this comment is confusing. "The seed most be incremented along" -> "The seed must be incremented after"? And if it's after the first 8 bytes, why is the index 16?
I would avoid talking about "incrementing the seed" because we're really incrementing a counter that exists alongside the seed.

Moreover, it seems like if we are going to treat part of `entropy` as a counter, we might as well do away with the `progress` field and pack both counters into `entropy`.

lukechampine commented on fastrand.go:
```diff
@@ -111,6 +108,9 @@ func Intn(n int) int {
 // BigIntn returns a uniform random value in [0,n). It panics if n <= 0.
+//
+// TODO: This particular function has not been extended to be faster, as it is
+// just a passthrough.
```
I don't think we can improve on `rand.Int`. It's already much faster (10x) since we pass in our own Reader instead of the crypto.rand Reader.
Benchmark:
```
BenchmarkBigIntn-4         	 3000000	       569 ns/op
BenchmarkBigIntnCrypto-4   	  300000	      4424 ns/op
```

lukechampine replied:
The `BenchmarkBigIntn` test uses a 32-byte big.Int, so its performance can be compared with `BenchmarkRead32` (which takes ~200ns). Profiling, it looks like the main reason `BigIntn` is slower is that it has to allocate the big.Int on the heap. If we were reading into a preallocated big.Int I imagine the benchmark would drop to nearly 200ns. Not worth the more annoying API, though.

DavidVorick replied:
Oh, I did not see that we are passing in our own reader. That's good enough then

lukechampine commented on fastrand.go:
```diff
@@ -20,65 +20,62 @@ import (
-	for i := 0; i < runtime.NumCPU(); i++ {
-		go r.fillEntropy()
+	r := &randReader{}
+	n, err := rand.Read(r.entropy[:])
```
following up on the other comment, I would make this `rand.Read(r.entropy[16:])` so that the counters are initialized to zero.

lukechampine commented on fastrand.go:
```diff
@@ -42,36 +49,52 @@ func init() {
+	// space.
+	if counter > 1<<63 {
+		atomic.AddUint64(&r.counterExtra, 1)
+		atomic.StoreUint64(&r.counter, 0)
```
is the `Store` necessary?

lukechampine commented on fastrand.go:
```diff
@@ -42,36 +49,52 @@ func init() {
+	// entropy within this thread.
 	n := 0
+	innerCounter := uint64(0)
+	innerCounterExtra := uint64(0)
```
prefer `var innerCounter, innerCounterExtra uint64`

The names are a bit verbose for my tastes -- I'd be fine with `c1`, `c2`, `c3`, `c4` -- but not worth fussing over.

DavidVorick replied:
But not `var n int`?

lukechampine replied:
I suppose. Mostly I just dislike using `:=` to initialize a zero-value.

lukechampine commented on fastrand.go:
```diff
@@ -42,36 +49,52 @@ func init() {
+	counterExtra := atomic.LoadUint64(&r.counterExtra)
+	// Update the second 64 bits of the counter if the first 64 bits are close
+	// to wrapping around. It is possible that the second 64 bits of the counter
+	// is updates multiple times by several concurrent threads. This wastes part
```
It is possible for the second 64 bits of the counter to be incremented multiple times by several concurrent threads.

lukechampine commented on fastrand.go:
```diff
@@ -42,36 +49,52 @@ func init() {
+	// space.
+	if counter > 1<<63 {
+		atomic.AddUint64(&r.counterExtra, 1)
+		atomic.StoreUint64(&r.counter, 0)
```
There is a potential overlap here:

1. thread 1 has `counter = 1<<63 + 1`, `counterExtra = 0`
2. thread 2 has `counter = 1<<63 + 2`, `counterExtra = 0`
3. thread 1 increments `counterExtra`
4. thread 1 resets `counter` to 0
5. thread 2 increments `counterExtra`
6. a new thread (3) calls `Read` and grabs `counter = 1`, `counterExtra = 2`
7. thread 2 resets `counter` to 0
8. a new thread (4) calls `Read` and grabs `counter = 1`, `counterExtra = 2`

lukechampine replied:
Would simply using `if counter == 1<<63` work? That way, it is guaranteed that only one thread will reset `r.counter`, and you still get overflow protection.

DavidVorick replied:
ah foo. you are right. Also I think the golang memory model does not guarantee the order in which things will happen within a thread, so I'm not sure you can even trust that the reset will happen after the increment.

hmm. I'd rather not just accept a 2^64 cycle though =/

DavidVorick replied:
Using `==` doesn't work because one thread could get `counter = 1<<63+1, extra = 1` when extra flips from 0 to 1 (because the thread grabs counter, and then extra increments, and then it grabs extra) and then on the next round you have a thread grabbing `counter = 1 << 63+1, extra =1` when the counter is flipping from 1 to 2. So you get overlaps, except this time they are extremely far apart and much less likely to be detected or disruptive.

But it's still an issue.

DavidVorick replied:
ok, I think the solution is actually not so bad.

Increment extra at `counter == 1<<63` and again at `counter == math.MaxUint64`. You have 2^63 time between each call, and it also guarantees that by the time it's wrapped all the way around back to the increment point, the counter is definitely higher than it's ever been before even if the threads are overlapping a bit.

lukechampine commented on fastrand.go:
```diff
@@ -42,36 +49,52 @@ func init() {
+	// Copy the counter and entropy into a separate slice, so that the result
+	// may be used in isolation of the other threads. The counter ensures that
+	// the result is unique to this thread.
+	seed := make([]byte, 64)
```
the slowdown you observed is probably because the seed is 24 bytes larger.

DavidVorick commented:
Added a test to check for repeat values in the entropy, but I don't think it'd be likely to find the mistake you described. I wrote it anyway.

Also updated both the comment and the strategy used to increment `counterExtra`, it should now be exceedingly unlikely that two different threads end up with the same values in the counters. But you should check the new strategy too in case there's yet again something I missed.

Also added some math/rand parallelism benchmarks. Interestingly enough, performance degrades a lot for math/rand when calling from multiple threads. Enough that fastrand will actually outperform it.

lukechampine commented:
The degradation is because the default `math/rand.Rand` object uses a mutex. It can't hope to compete with atomic operations.

lukechampine commented on fastrand.go:
```diff
@@ -53,16 +53,21 @@ func (r *randReader) Read(b []byte) (int, error) {
+	//
+	// Note that there could still be a problem if 2^63 threads all manage to
+	// increment 'counter' before counterExtra is incremented by the first
+	// thread. This has been deemend to be sufficiently unlikely.
```
The code looks good but the comment doesn't explain it super well. Granted, it's not an easy thing to explain concisely. Here's my stab at it:
```
// Increment counterExtra when counter is close to overflowing. We cannot wait
// until counter == math.MaxUint64 to increment counterExtra, because another
// goroutine could call Read, overflowing counter to 0 before the first
// goroutine increments counterExtra. The second goroutine would then be
// reusing the counter pair (0, 0). Instead, we increment at 1<<63 so that
// there is little risk of an overflow.
//
// There is still a potential overlap near 1<<63, though, because another
// goroutine could see counter == 1<<63+1 before the first goroutine
// increments counterExtra. The counter pair (1<<63+1, 1) would then be
// reused. To prevent this, we also increment at math.MaxUint64. This means
// that in order for an overlap to occur, 1<<63 goroutine would need to
// increment counter before the first goroutine increments counterExtra.
```
Also, I think the branch could be:
```go
if counter == 1<<63 || counter == math.MaxUint64 {
```
Since we're already assuming that `counter` won't be incremented `1<<63` times during a single goroutine

DavidVorick commented:
ok, updated.
